content/news/2023-06-from-single-files-to-collections/index.md

```diff
@@ -0,0 +1,115 @@
+
+# Case 2 - Failing or empty elements in a collection
+
+Even if a workflow is well designed, in some cases in can happen that only few elements of a collection fail. This happened to us rather randomly in case of Kraken2, since
+it requires large amounts of memory (>70 GB), which were not assigned to every run of the tool by the server. That issue was solved by increasing the minimum memory required by the tool on the EU server (`TODO how was that done`). But there are various other scenarios where the failure of the tool can be attributed e.g. to specific input data. In other cases only a few elements of a collection are empty (e.g. if an assembly can not be made due to not overlapping reads).
+
+If an element of a collection is failed or empty the entire downstream processing is stopped, which can be rather annoying if one want to process large amount of data and got stuck due 
+to a few elements. Two solutions are proposed to handle such cases.
+
+## Intermediate workflow specific solution
+
+Collections can be filtered for failed or empty datasets using collection tools such as [filter empty datasets](https://usegalaxy.eu/?tool_id=__FILTER_EMPTY_DATASETS__&version=1.0.0)  
+and [filter failed datasets](https://usegalaxy.eu/?tool_id=__FILTER_FAILED_DATASETS__&version=1.0.0). 
+
+<div class="center">
+<div class="img-sizer" style="width: 100%">
+
+![Empty collection](./figs/empty_collection.png)
+
+</div>  
+<figcaption>
+  Filter failed collection elements
+</figcaption>
+</div>
+
+<div class="center">
+<div class="img-sizer" style="width: 100%">
+
+![Empty collection](./figs/failed_collection.png)
+
+</div>  
+<figcaption>
+  Filter empty collection elements
+</figcaption>
+</div>
+
+
+Although this can solve the issue in some cases immediately, further considerations need to be made. 
+First, often one cannot really know at which step the collection will be affected.
+To cover all cases one would need to add the filter steps for every produced collection, which will increase the workflow steps unreasonably. 
+Secondly, the filtering will change the size of the collection. If downstream tools depend on a specific collection size, which is always the case if a tool
+takes two or more collections as input, the tool will also fail. That's basically a follow up problem of the first problem.
+This case can still be solved by a intermediate step where the second collection is basically reindexed by the same element identifiers then the collection with
+missing elements. 
+
+<div class="center">
+<div class="img-sizer" style="width: 100%">
+
+![Empty collection](./figs/reindex_collection.png)
+
+</div>  
+<figcaption>
+  Proposed solution to reindex second collection using elements of the first collection
+</figcaption>
+</div>  
+
+However, if this logic needs to be applied to multiple collection or again for multiple steps, the workflow becomes even more unreasonably large and complex. 
```

Comment:
You can also run the workflow again and enable the job cache, this might be the easiest and most clean solution.

Reply:
@mvdbeek do you mean to run the workflow again without the failed elements ? Then if another collection fails, run the complete WF again and so on ? How do you enable the job cache and what does it do ?

Reply:
No, just run the whole workflow again. Everything that had already been run successfully will be skipped. If enabled by the admin go to User > Prefernces -> Manage Information. There were some additional fixes in 23.1 and usegalaxy.eu isn't running this yet. If you want to test this you can try on usegalaxy.org.

Reply:
Thank you for this info, cached jobs are indeed really cool, but I don't really see how it can solve the current issue.
The same elements will still fail again. Only in the case of randomly failing, this could help ... which should not be the case in any way.

Therefore, an additional step would still be required, by:
* Filtering the elements in the collection (-> reindexing problem)
* Run WF again without the Datasets that fail (-> laborious)
* Work on the tool to overcome the issue

Or am I missing something ?

Reply:
Your example said

> Even if a workflow is well designed, in some cases in can happen that only few elements of a collection fail. This happened to us rather randomly in case of Kraken2, since
it requires large amounts of memory (>70 GB), which were not assigned to every run of the tool by the server.

so it should help wiht that ?

Reply:
Ah ok, sorry I thought you referred to the complete paragraph, for that particular example your solution can help, although since it's random, one could get some failed ones for the new run as well and ultimately would need to repeat the workflow multiple times, which is also not ideal

Reply:
No, those will be picked up. And it's generally a good thing if you have a single complete invocation without partially failed items.